contributors/design-proposals/minreadyseconds.md

```diff
@@ -0,0 +1,73 @@
+
+Two problems that have been identified so far with the current state of things:
+1. A Pod is marked ready by the kubelet as soon as it passes its readiness check. The
+ReplicaSet controller runs as part of master and estimates when a Pod is available by
+comparing the time the Pod became ready (as seen by the kubelet) with MinReadySeconds.
+[Clock-skew between master and nodes will affect the availability checks](https://github.com/kubernetes/kubernetes/issues/29229). 
+2. PodDisruptionBudget is working with ready Pods and has no notion of MinReadySeconds
+when used by a Deployment/ReplicaSet. 
+
+Both problems above can be solved by moving MinReadySeconds in the PodSpec. Once kubelet
+observes that a Pod has been ready for at least MinReadySeconds without any of its containers
+crashing, it will update the PodStatus with an Available condition set to Status=True.
+Higher-level orchestrators running on different machines such as the ReplicaSet or the
+PodDisruptionBudget controller will merely need to look at the Available condition that
+is set in the status of a Pod.
+
+## API changes
+
+A new field is proposed in the PodSpec:
+```go
+	// Minimum number of seconds for which a newly created pod should be ready
+	// without any of its container crashing, for it to be considered available.
+	// Defaults to 0 (pod will be considered available as soon as it is ready)
+	// +optional
+	MinReadySeconds *int32 `json:"minReadySeconds,omitempty"`
+```
+and a new PodConditionType:
+```go
+	// PodAvailable is added in a ready pod that has MinReadySeconds specified. The pod
+	// should already be added under a load balancer and serve requests, this condition
+	// lets higher-level orchestrators know that the pod is running after MinReadySeconds
+	// without having any of its containers crashed.
+	PodAvailable PodConditionType = "Available"
+```
+Additionally:
+* Deployments/ReplicaSets/DaemonSets already use MinReadySeconds in their spec so we
+should probably deprecate those fields in favor of the field in the PodSpec and remove
+them in a future version. 
+* Deployments/ReplicaSets will not propagate MinReadySeconds from their Spec down to the
+pod template because that will lead in differences in the pod templates between a
+Deployment and a ReplicaSet resulting in new rollouts. If MinReadySeconds is specified
+both in the spec and pod template for a Deployment/ReplicaSet and it's not the same value,
+a validation error will be returned (tentative). API defaulting can set MinReadySeconds
+in the spec, if it's specified only in the PodTemplate (tentative).
+* ReplicaSets that specify MinReadySeconds only in the ReplicaSetSpec, can create new Pods
+by specifying MinReadySeconds in their PodSpec (w/o updating the ReplicaSet pod template).
+
+## kubelet changes
+For a Pod tha specifies MinReadySeconds, kubelet will need to check (after MinReadySeconds)
+if any of the Pod containers has crashed. If not, it will switch the Available condition to
```

Comment:
So there seem to be two types of errors an application can have thus a 500 that a user will see: 1) those that will make the application exit. In this case, the pod goes out of rotation anyway, it never transitions to Stable, the Deployment blocks. We can simply verify these by looking at the container statuses. 2) Bugs that may not cause the application to exit. Pod is not going out of rotation. Deployment still blocks. Can't look at container statuses, readiness checks may not be enough unless business metrics are involved.

Both cases are covered by rerunning the readiness check I guess. I need to call this out. Does it also mean that the endpoints controller will need to remove endpoints for Pods that never reach Stable?

6 does not seem to be something that needs to be handled in this proposal. We have ProgressDeadlineSeconds and there is AutoRollback that needs to be implemented.